framework/di/Container.php

```diff
@@ -456,48 +457,68 @@ protected function resolveDependencies($dependencies, $reflection = null)
+            $reflection = new \ReflectionFunction($callback);
+        }
 
-            $args = [];
-            foreach ($reflection->getParameters() as $param) {
-                $name = $param->getName();
-                if (($class = $param->getClass()) !== null) {
-                    $className = $class->getName();
-                    if (isset($params[0]) && $params[0] instanceof $className) {
-                        $args[] = array_shift($params);
-                    } elseif (\Yii::$app->has($name) && ($obj = \Yii::$app->get($name)) instanceof $className) {
-                        $args[] = $obj;
-                    } else {
-                        $args[] = $this->get($className);
-                    }
-                } elseif (count($params)) {
+        $args = [];
+
+        $associative = ArrayHelper::isAssociative($params);
+
+        foreach ($reflection->getParameters() as $param) {
+            $name = $param->getName();
+            if (($class = $param->getClass()) !== null) {
+                $className = $class->getName();
+                if ($associative && isset($params[$name]) && $params[$name] instanceof $className) {
+                    $args[] = $params[$name];
+                } elseif (!$associative && isset($params[0]) && $params[0] instanceof $className) {
                     $args[] = array_shift($params);
-                } elseif ($param->isDefaultValueAvailable()) {
-                    $args[] = $param->getDefaultValue();
-                } elseif (!$param->isOptional()) {
-                    $funcName = $reflection->getName();
-                    throw new InvalidConfigException("Missing required parameter \"$name\" when calling \"$funcName\".");
+                } elseif (\Yii::$app->has($name) && ($obj = \Yii::$app->get($name)) instanceof $className) {
+                    $args[] = $obj;
+                } else {
+                    $args[] = $this->get($className);
                 }
+            } elseif ($associative && isset($params[$name])) {
+                $args[] = $params[$name];
+            } elseif (!$associative && count($params)) {
+                $args[] = array_shift($params);
+            } elseif ($param->isDefaultValueAvailable()) {
+                $args[] = $param->getDefaultValue();
+            } elseif (!$param->isOptional()) {
+                $funcName = $reflection->getName();
+                throw new InvalidConfigException("Missing required parameter \"$name\" when calling \"$funcName\".");
             }
+        }
 
-            foreach ($params as $value) {
-                $args[] = $value;
-            }
-            return call_user_func_array($callback, $args);
+        foreach ($params as $value) {
+            $args[] = $value;
+        }
+        return $args;
+    }
+
+    /**
+     * Invoke callback with resolved dependecies parameters.
+     * @param callable $callback callable to be invoked.
+     * @param array $params callback paramater.
```

Comment:
also need more documentation about this parameter works. from the code above it looks like named params are allowed. should be mentioned here.